Indexing/SubstitutionTree.hpp

```diff
@@ -108,200 +108,193 @@ using namespace Kernel;
-   */
-  class SubstitutionTree
-  {
-  public:
-    SubstitutionTree() = default;
-    SubstitutionTree(SubstitutionTree const&) = delete;
-    SubstitutionTree& operator=(SubstitutionTree const& other) = delete;
-    SubstitutionTree(SubstitutionTree&& other)
-    : SubstitutionTree()
-    {
-      std::swap(_nextVar, other._nextVar);
-      std::swap(_root, other._root);
-    }
+/**
+ * Class of substitution trees. 
+ *
+ * We can either store typed terms, or literals in a subtitution tree.
+ * Classically we'd think of inserting/removing only one term t into a substitution tree. 
+ * This can be understood as inserting the substitution { S0 -> t } into the tree.
+ *
+ * In general we can insertt a substitution with more than just this one binding. 
+ * This is what we do in order to store the sort of variables, and in order to insert all the arguments of a literal:
+ * - For a term t of sort s we insert { S0 -> t; S1 -> s }
+ * - For literals (~)P(t0..tn) we insert { S0 -> t0 .. Sn -> tn }.
+ * (Note that we do not check the predicate or the polarity of literals here. This happens in LiteralSubstitutionTree)
+ */
+template<class LeafData_>
+class SubstitutionTree final
+{
 
-    virtual ~SubstitutionTree();
+public:
+  using LeafData = LeafData_;
 
-    friend std::ostream& operator<<(std::ostream& out, SubstitutionTree const& self);
-    friend std::ostream& operator<<(std::ostream& out, OutputMultiline<SubstitutionTree> const& self);
+  static constexpr int QRS_QUERY_BANK = 0;
+  static constexpr int QRS_RESULT_BANK = 1;
 
-    struct LeafData {
-      LeafData() {}
 
-      LeafData(Clause* cls, Literal* literal, TypedTermList term, TermList extraTerm)
-      : clause(cls), literal(literal), term(term), sort(term.sort()), extraTerm(extraTerm) {}
-      LeafData(Clause* cls, Literal* literal, TypedTermList term)
-      : clause(cls), literal(literal), term(term), sort(term.sort()), extraTerm(TermList::empty()) {}
+  SubstitutionTree(unsigned reservedSpecialVars)
+    : _nextVar(reservedSpecialVars)
+    , _root(nullptr)
+  {}
 
-      LeafData(Clause* cls, Literal* literal, TermList term, TermList extraTerm)
-      : clause(cls), literal(literal), term(term), sort(TermList::empty()), extraTerm(extraTerm) {}
-      LeafData(Clause* cls, Literal* literal, TermList term)
-      : clause(cls), literal(literal), term(term), sort(TermList::empty()), extraTerm(TermList::empty()) {}
+  SubstitutionTree(SubstitutionTree const&) = delete;
+  SubstitutionTree& operator=(SubstitutionTree const& other) = delete;
+  static void swap(SubstitutionTree& self, SubstitutionTree& other) {
+    std::swap(self._nextVar, other._nextVar);
+    std::swap(self._root,    other._root);
+  }
+  SubstitutionTree& operator=(SubstitutionTree && other) { swap(*this,other); return *this; }
+  SubstitutionTree(SubstitutionTree&& other) : SubstitutionTree(0) { swap(*this, other); }
```

Comment:
Copy-and-swap idiom but only for move and move-assign, right?

Reply:
Yes. But could we use it for anything else as well?

Reply:
In principle also for copy/copy-assign, but I don't think we want to. Just checking this.